File: modules/supabase-core-typescript/src/utils/cache.ts
/**
 * Cache Utilities
 * 
 * Simple in-memory caching for Supabase query results.
 * For production use, consider Redis or a more robust caching solution.
 * 
 * Note: This is a basic implementation. For production, use a proper cache like Redis.
 */

/**
 * Cache entry with expiration.
 */
interface CacheEntry<T> {
  data: T
  expiresAt: number
}

/**
 * Simple in-memory cache for Supabase query results.
 * 
 * @example
 * ```typescript
 * const cache = new QueryCache()
 * 
 * // Cache a query result
 * const key = 'posts:published'
 * const { data } = await supabase.from('posts').select('*').eq('published', true)
 * cache.set(key, data, 300) // Cache for 5 minutes
 * 
 * // Retrieve from cache
 * const cached = cache.get(key)
 * if (cached) {
 *   return cached
 * }
 * ```
 */
export class QueryCache {
  private cache: Map<string, CacheEntry<any>> = new Map()
  private defaultTTL: number = 300 // 5 minutes

  /**
   * Sets a cache entry.
   * 
   * @param key - Cache key
   * @param data - Data to cache
   * @param ttlSeconds - Time to live in seconds (default: 5 minutes)
   */
  set<T>(key: string, data: T, ttlSeconds?: number): void {
    const ttl = ttlSeconds || this.defaultTTL
    const expiresAt = Date.now() + ttl * 1000

    this.cache.set(key, {
      data,
      expiresAt,
    })
  }

  /**
   * Gets a cache entry.
   * Returns null if not found or expired.
   * 
   * @param key - Cache key
   * @returns Cached data or null
   */
  get<T>(key: string): T | null {
    const entry = this.cache.get(key)

    if (!entry) {
      return null
    }

    // Check if expired
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key)
      return null
    }

    return entry.data as T
  }

  /**
   * Deletes a cache entry.
   * 
   * @param key - Cache key
   */
  delete(key: string): void {
    this.cache.delete(key)
  }

  /**
   * Clears all cache entries.
   */
  clear(): void {
    this.cache.clear()
  }

  /**
   * Cleans up expired entries.
   * Call this periodically to free memory.
   */
  cleanup(): void {
    const now = Date.now()
    for (const [key, entry] of this.cache.entries()) {
      if (now > entry.expiresAt) {
        this.cache.delete(key)
      }
    }
  }

  /**
   * Gets cache statistics.
   * 
   * @returns Cache stats
   */
  getStats(): {
    size: number
    entries: number
  } {
    this.cleanup() // Clean up before counting
    return {
      size: this.cache.size,
      entries: this.cache.size,
    }
  }
}

/**
 * Global cache instance.
 * Use this for simple caching needs.
 */
export const globalCache = new QueryCache()

/**
 * Creates a cache key from query parameters.
 * Useful for generating consistent cache keys.
 * 
 * @param table - Table name
 * @param params - Query parameters
 * @returns Cache key string
 * 
 * @example
 * ```typescript
 * const key = createCacheKey('posts', { published: true, limit: 10 })
 * // Returns: 'posts:published=true:limit=10'
 * ```
 */
export function createCacheKey(table: string, params: Record<string, any>): string {
  const sortedParams = Object.keys(params)
    .sort()
    .map((key) => `${key}=${JSON.stringify(params[key])}`)
    .join(':')

  return `${table}:${sortedParams}`
}

File: modules/supabase-core-python/src/utils/cache.py
"""
Cache Utilities

Utilities for caching Supabase query results.
"""

from typing import Any, Optional, Dict, Callable
import hashlib
import json
import time


class QueryCache:
    """
    Simple in-memory cache for Supabase query results.

    Example:
        ```python
        from supabase_core_python import QueryCache

        cache = QueryCache(ttl_seconds=300)  # 5 minute TTL

        def fetch_posts():
            return supabase.table("posts").select("*").execute()

        # Cache the result
        result = cache.get_or_set("posts:all", fetch_posts)
        ```
    """

    def __init__(self, ttl_seconds: int = 300):
        """
        Initialize cache.

        Args:
            ttl_seconds: Time-to-live for cached entries (default: 300 seconds)
        """
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.ttl_seconds = ttl_seconds

    def get(self, key: str) -> Optional[Any]:
        """
        Get cached value.

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found/expired
        """
        if key not in self.cache:
            return None

        entry = self.cache[key]
        if time.time() > entry["expires_at"]:
            del self.cache[key]
            return None

        return entry["value"]

    def set(self, key: str, value: Any) -> None:
        """
        Set cached value.

        Args:
            key: Cache key
            value: Value to cache
        """
        self.cache[key] = {
            "value": value,
            "expires_at": time.time() + self.ttl_seconds,
        }

    def get_or_set(self, key: str, operation: Callable[[], Any]) -> Any:
        """
        Get cached value or execute operation and cache result.

        Args:
            key: Cache key
            operation: Function to execute if cache miss

        Returns:
            Cached or newly computed value
        """
        cached = self.get(key)
        if cached is not None:
            return cached

        result = operation()
        self.set(key, result)
        return result

    def clear(self) -> None:
        """Clear all cached entries."""
        self.cache.clear()

    def invalidate(self, key: str) -> None:
        """
        Invalidate a specific cache entry.

        Args:
            key: Cache key to invalidate
        """
        if key in self.cache:
            del self.cache[key]


# Global cache instance
global_cache = QueryCache()


def create_cache_key(*args: Any, **kwargs: Any) -> str:
    """
    Creates a cache key from arguments.

    Args:
        *args: Positional arguments
        **kwargs: Keyword arguments

    Returns:
        Cache key string

    Example:
        ```python
        from supabase_core_python import create_cache_key

        key = create_cache_key("posts", user_id=123, page=1)
        # Returns: "posts:user_id=123:page=1"
        ```
    """
    parts = []
    for arg in args:
        parts.append(str(arg))
    for key, value in sorted(kwargs.items()):
        parts.append(f"{key}={value}")

    # Create hash for long keys
    key_string = ":".join(parts)
    if len(key_string) > 200:
        key_string = hashlib.md5(key_string.encode()).hexdigest()

    return key_string

